vortex-array/src/array/chunked/flatten.rs

```diff
@@ -0,0 +1,146 @@
+        .collect::<VortexResult<Vec<_>>>()?;
+
+    let len = chunks.iter().map(|chunk| chunk.len()).sum();
+    let validity = chunks.iter()
+        .map(|chunk| chunk.logical_validity())
+        .collect::<Validity>();
+
+    let mut field_arrays = Vec::new();
+    let field_names = struct_dtype.names().clone();
+    let field_dtypes = struct_dtype.dtypes().clone();
+
+    for (field_idx, field_dtype) in field_dtypes.iter().enumerate() {
+        let mut field_chunks = Vec::new();
+        for chunk in &chunks {
+            field_chunks.push(chunk.field(field_idx).expect("structarray should contain field"));
+        }
+        let field_array = ChunkedArray::try_new(field_chunks, field_dtype.clone())?;
+        field_arrays.push(field_array.into_array());
+    }
+
+    Ok(StructArray::try_new(field_names, field_arrays, len, validity)?)
+}
+
+fn pack_bools(chunks: &[Array], nullability: Nullability) -> VortexResult<BoolArray> {
+    let len = chunks.iter().map(|chunk| chunk.len()).sum();
+    let mut logical_validities = Vec::new();
+    let mut bools = Vec::with_capacity(len);
+    for chunk in chunks {
+        let chunk = chunk.clone().flatten_bool()?;
+        logical_validities.push(chunk.logical_validity());
+        bools.extend(chunk.boolean_buffer().iter());
+    }
+
+    BoolArray::try_new(
+        BooleanBuffer::from(bools),
+        validity_from_chunks(logical_validities, nullability),
+    )
+}
+
+fn pack_primitives(chunks: &[Array], ptype: PType, nullability: Nullability) -> VortexResult<PrimitiveArray> {
+    let len: usize = chunks.iter().map(|chunk| chunk.len()).sum();
+    let mut logical_validities = Vec::new();
+    let mut buffer = MutableBuffer::with_capacity(len * ptype.byte_width());
+    for chunk in chunks {
+        let chunk = chunk.clone().flatten_primitive()?;
+        logical_validities.push(chunk.logical_validity());
+        buffer.extend_from_slice(chunk.buffer());
+    }
+
+    match_each_native_ptype!(ptype, |$T| {
+        Ok(PrimitiveArray::try_new(
+            ScalarBuffer::<$T>::from(buffer),
+            validity_from_chunks(logical_validities, nullability))?)
+    })
+}
+
+// TODO(aduffy): This can be slow for really large arrays.
+// TODO(aduffy): this doesn't propagate the validity fully
+fn pack_varbin(chunks: &[Array], dtype: DType, _nullability: Nullability) -> VortexResult<VarBinArray> {
+    let len = chunks.iter()
+        .map(|chunk| chunk.len())
+        .sum();
+    let mut builder = VarBinBuilder::<i32>::with_capacity(len);
+
+    for chunk in chunks {
+        let chunk = chunk.clone().flatten_varbin()?;
+        chunk.with_iterator(|iter| {
+            for datum in iter {
+                builder.push(datum);
+            }
+        })?;
+    }
+
+    Ok(builder.finish(dtype))
+}
+
+fn validity_from_chunks(logical_validities: Vec<LogicalValidity>, nullability: Nullability) -> Validity {
+    if nullability == Nullability::NonNullable {
```

Comment:
Shouldn't this case be handled by the collect implementation?

Reply:
so this is something i continue to be confused by w.r.t. Validity VS LogicalValidity.

LogicalValidity doesn't have a `NonNullable` variant, but Validity does. Because of this, the collect implementation cannot return `Validity::NonNullable`.

Most of the canonical encodings did this in their impl of AsContiguousFn, e.g. here's the one for PrimitiveArray: https://github.com/spiraldb/vortex/blob/3308c6c07d2ea8bf2a0d70922befd75ba95c39a3/vortex-array/src/array/primitive/compute/as_contiguous.rs#L13-L17

Reply:
The idea in those implementations I think is to avoid _generating_ the logical validity if it can be avoided.

Since we already have them (at least in this function), the collect implementation actually collects from LogicalValidity into Validity: `impl FromIterator<LogicalValidity> for Validity `

Reply:
Fixed this implementation to lazily compute the logical_validities rather than doing it eagerly.

Re the collection implementation: you can get back a Validity::AllValid(usize) from the collector, but you can't get back a Validity::NonNullable, so you'd still need this if/else regardless